### PR TITLE
[Product] ProductFactory with variant setting in createNew()

### PR DESCRIPTION
**Pull request by TheMadeleine**
| Q                    | A
| -------------       | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Fixed tickets | #480 
| License       | MIT

### CODE REVIEW

File: src/Sylius/Component/Product/Factory/ProductFactory.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+     * @param string $className
+     */
+    public function __construct(
+        FactoryInterface $translatableResourceFactory,
+        LocaleProviderInterface $localeProvider,
+        FactoryInterface $variantFactory,
+        $className
+    ) {
+        $this->variantFactory = $variantFactory;
+        parent::__construct($translatableResourceFactory, $localeProvider);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Product/Factory/ProductFactory.php]: `parent` should be first call in constructor.

File: src/Sylius/Component/Product/spec/Factory/ProductFactorySpec.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Sylius\Component\Product\Factory\ProductFactory');
+    }
+
+    function it_implements_factory_interface()
+    {
+        $this->shouldImplement('Sylius\Component\Resource\Factory\FactoryInterface');
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Product/spec/Factory/ProductFactorySpec.php]: Please use `FactoryInterface::class` here.

File: src/Sylius/Component/Product/spec/Factory/ProductFactorySpec.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
+use Sylius\Component\Translation\Provider\LocaleProviderInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+
+class ProductFactorySpec extends ObjectBehavior
+{
+    function let(FactoryInterface $variantFactory, FactoryInterface $factory, LocaleProviderInterface $localeProvider)
+    {
+        $this->beConstructedWith($factory, $localeProvider, $variantFactory, 'Sylius\Component\Product\Model\Variant');
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Product/spec/Factory/ProductFactorySpec.php]: What is the last argument here?

File: src/Sylius/Bundle/ProductBundle/DependencyInjection/SyliusProductExtension.php [hunk 4]
```diff
@@ -57,6 +59,12 @@ public function load(array $config, ContainerBuilder $container)
         foreach ($configFiles as $configFile) {
             $loader->load($configFile);
         }
+
+        $container
+            ->getDefinition('sylius.factory.product')
+            ->addArgument(new Reference('sylius.factory.product_variant'))
+            ->addArgument(new Parameter('sylius.model.product_variant.class'))
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ProductBundle/DependencyInjection/SyliusProductExtension.php]: I suppose it's not needed.

**Comment by pjedrzejewski** [comment]
So happy to get rid of this issue, thanks Magda! :+1:

**Pull request merged**